File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        vector<int> orig;
        int n = s.size();
        for(int i=0;i<n;i++){
            char c = s[i];
            if(c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' || c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U')orig.push_back(i);
        }
        vector<int> change = orig; reverse(change.begin(), change.end());
        for(int i=0;i<change.size()>>1;i++)swap(s[orig[i]],s[change[i]]);
        return s;
    }
};

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        st = {'a','e','i', 'o','u'}
        mp = []
        s = list(s)
        for i in range(len(s)):
            if s[i].lower() in st:
                mp.append([s[i], i])

        mp.reverse()
        
        i,j =0, len(mp) - 1

        while i < j:
            s[mp[i][1]], s[mp[j][1]] = s[mp[j][1]],s[mp[i][1]]
            i+=1
            j-=1
        return ''.join(s)